### DISCUSSION TITLE
Making CTRL for Prostage Remote Wah, stuck on 1 parameter

### DISCUSSION THREAD

**Discussion by verchrona** [RomanKubiak/ctrlr#657] [General]
Hello,

In the Prostage Remote Wah manual, there is a wah to edit the Auto Wah tempo, but this requires Sysex instead of CCs. I've already got the other parameters mapped. This one is giving me a bit of trouble.

(FROM MANUAL)
========
Auto WahWah Speed:
F0 00 20 75 00 76 XX F7
F0 00 20 75 00 77 YY F7 
Auto WahWah speed (Time in ms = 2 * XXYY)
Warning: The entry is made in HEX! E.g. "0100" corresponds to binary 128, and the resulting time is therefore 2 * 128 = 256 ms)
========

This part confused the heck out of me because, isn't 10000000 = 128? How does 0100 = 128 from a binary set of numbers according to the manual? in

From the Midi Binary information, 0100 = 4

To get 128 in HEX, that would be 80, correct? I'm not sure how that is spread over 2 different message lines though.

Is this something that can be accomplished with the built in Sysex Multi message or does this require LUA scripting?

**Comment by damiensellier** [comment]
10000000 is binary 8 digit number (forget about it) 
128 in "decimal" is 100 in hexadecimal. I guess they made a mistake on the denomination.


xx is the 1st byte (7bit) and yy being the 2nd  byte (7bit) 

so for a decimal value of 128 the message is : 

F0 00 20 75 00 76 00 F7 (msb) with XX = 00h
F0 00 20 75 00 77 80 F7 (lsb) with YY = 80h

I'm pretty sure it's just sending MSB and LSB via sysex, knowing the value sent has to be devided by 2 beforehand.


![Capture d’écran, le 2024-06-27 à 02 47 05](https://github.com/RomanKubiak/ctrlr/assets/55164385/bdc23ed9-1d06-4ffb-aa07-57df672252b0)

The calculator:
![Capture d’écran, le 2024-06-27 à 02 46 58](https://github.com/RomanKubiak/ctrlr/assets/55164385/99f63a5e-3b6e-4304-b465-87f252102989)

The multi message settings:
![Capture d’écran, le 2024-06-27 à 02 46 48](https://github.com/RomanKubiak/ctrlr/assets/55164385/7842e6f7-1401-436e-98d3-dc7a6baba2d5)

You'll get this kind of message : 
![Capture d’écran, le 2024-06-27 à 02 46 20](https://github.com/RomanKubiak/ctrlr/assets/55164385/3c5cc1d9-6f50-412a-b9f1-c6ca7398d1cc)

Sample panel:

![Capture d’écran, le 2024-06-27 à 02 50 34](https://github.com/RomanKubiak/ctrlr/assets/55164385/66fcd1fd-be0b-4922-9540-483e5332e3af)

According to the manual it goes from 40ms to 5000ms.

See attached panel if it works

Peace!

Damien


[Remote WahWah Speed Sysex_1_0_2024-06-27_02-47.panel.zip](https://github.com/user-attachments/files/16009272/Remote.WahWah.Speed.Sysex_1_0_2024-06-27_02-47.panel.zip)

**Reply by damiensellier** [reply]
I invite you to take a look at Spiffo's video on youtube about sending SysEx messages from Ctrlr: 
https://www.youtube.com/watch?v=RPp_7Ht4-UM&t=237s

Enjoy ;)

**Reply by damiensellier** [reply]
Also for info here is the list of different tokens with sysex messages (xx, y,ls,LS, ms,MS, r1,z5 etc) : 

https://github.com/RomanKubiak/ctrlr/blob/8aa00d82127acda42ad9ac9b7b479461e9436aa4/Source/Core/CtrlrSysexProcessor.cpp#L281

```
if (s == "xx"){return (ByteValue);}
if (s.endsWith("y") && !s.startsWith("y")){return (ByteChannel4Bit);}
if (s.startsWith("k") || s.startsWith("p") || s.startsWith("n") || s.startsWith("o")){return (GlobalVariable);}
if (s.startsWith("z")){return (ChecksumRolandJP8080);}
if (s.startsWith("w")){return (ChecksumWaldorfRackAttack);}
if (s.startsWith("u")){return (LUAToken);}
if (s.startsWith ("v")){return (FormulaToken);}
if (s == "yy"){return (ByteChannel);}
if (s == "LS"){return (LSB7bitValue);}
if (s == "MS"){return (MSB7bitValue);}
if (s == "ii"){return (Ignore);}
if (s == "ls"){return (LSB4bitValue);}
if (s == "ms"){return (MSB4bitValue);}
if (s == "r1"){return (RolandSplitByte1);}
if (s == "r2"){return (RolandSplitByte2);}
if (s == "r3"){return (RolandSplitByte3);}
if (s == "r4"){return (RolandSplitByte4);}
if (s == "tp"){return (CurrentProgram);}
if (s == "tb"){return (CurrentBank);}
```

**Reply by verchrona** [reply]
Ah thanks so much for making that Panel demo. I will try this tomorrow with the unit!

In the meantime, I have been trying to educate myself on how you arrived at the equation values. For the last 2 hours, I have been staring at the equation "Time in ms = 2 * XXYY" and trying so hard to figure out how the modulator values apply.

For instance:
When the panel is set at 128 ms, we get 00 and 40 as the XX and YY.
So 128ms = 2 * 64

That equation makes sense to me so far.

When the panel is set at 254 ms, we get 00 and 7f as the XX and YY.
So 254ms = 2 * 127

This also makes sense to me.

Here is what trips me up.
When the panel is set at 256 ms, we get 01 and 00 as the XX and YY.
What happened to all the Hex values after 7f? The values 80 to FF are skipped and it adds 1 to the XX group. Shouldn't the XXYY for 256 ms be 0080?
So 256ms does not equal 2*256.
If XXYY for 256 ms was 0080, then 256ms = 2 * 128

I see the same pattern repeat every 256 ms.
256ms --> XXYY = 0100
512ms --> XXYY = 0200
768ms --> XXYY = 0300

After 256 ms, it no longer multiplies by 2? Or was is supposed to only multiply the YY by 2, and not XX?

**Reply by damiensellier** [reply]
It's called byte spliting with MSB and LSB. Most significant bits, Less Significant bits.

MIDI is 7bits bytes, always.
byte values goes from 0 to 127
usually for wider range of values we split the value on 2 bytes MSB & LSB, here XX and YY/
It allows to get value range up to 16384 (128 x 128) which means 0 to 16383. That's a major improvement.

from 0 to 127 you have YY (LSB) = 0 to 7F and of course XX = 0 (MSB).
for values above 127 you add 1 to XX everytime you reach a lsb YY of 128(0) .


The x2 formula is something they decided probably for precision purpose. From the manual they just want you to divide 2 times the whole time value for the message value.   256 ms has a value of (128)dec, 1s has a value of (500)dec

**Reply by damiensellier** [reply]
https://www.morningstar.io/post/2016/12/25/midi-msb-and-lsb

**Reply by damiensellier** [reply]
To be clear with an exemple : 

For exemple: 
a value of 5000ms would be 5000/2 = 2500
floor(2500/128) = 19
knowing that 19x128 = 2432
So 2500-2432 = 68 (remainder)

Here: 
MSB is 19dec | 13hex
LSB is 68dec | 44hex

message would be : 

F0 00 20 75 00 76 13 F7
F0 00 20 75 00 77 44 F7


![Capture d’écran, le 2024-06-27 à 15 27 57](https://github.com/RomanKubiak/ctrlr/assets/55164385/99ca2a2e-5289-4d69-b185-1cd073a4ec0e)


Hope it helps

Damien

**Reply by verchrona** [reply]
Thanks very much for explaining that to me! I just got my intro to Hexadecimal yesterday. The control works perfectly with my Prostage Wah unit! I realized this unit may have a slower processor, so typing in the ms value definitely helps it not choke on the fast midi sysex output.

That example above explains the equation very well, I was scratching my head for hours yesterday trying to unravel it. Wonderful work coding the control and explaining the process clearly, Damien! Very much appreciated!

**Reply by damiensellier** [reply]
Thanks, you're welcome! happy it's working from the get go. 
You can slow down the stream of data sent to the pedal with the timer method (sending 1 value every 20ms for exemple) or just send the value only when you release the mouse button. "send value only on release"

![Capture d’écran, le 2024-06-28 à 02 11 54](https://github.com/RomanKubiak/ctrlr/assets/55164385/a892ed19-bb16-4129-9a35-4129001ccb9c)

**Reply by verchrona** [reply]
Ah great tip! If only I knew that earlier, I returned a guitar pedal a while back specifically because it couldn't keep up with fast values!

**Reply by damiensellier** [reply]
It's a new feature we added a few months ago only for that particular purpose. It's pretty useful for old synths too.

**Comment by dnaldoog** [comment]
> Hello,
> 
> In the Prostage Remote Wah manual, there is a wah to edit the Auto Wah tempo, but this requires Sysex instead of CCs. I've already got the other parameters mapped. This one is giving me a bit of trouble.
> 
> # (FROM MANUAL)
> # Auto WahWah Speed:
> F0 00 20 75 00 76 XX F7
> F0 00 20 75 00 77 YY F7
> Auto WahWah speed (Time in ms = 2 * XXYY)
> Warning: The entry is made in HEX! E.g. "0100" corresponds to binary 128, and the resulting time is therefore 2 * 128 = 256 msm

It looks to me like they are sending two messages per value.  The first XX is the msb; YY is the lsb. Hex 80 is 128 but one sysex data byte has a limit of 7f (127) so to send a value over 127 you need two bytes 01 * 128 = 128 XX and 00 = 0 sent in two messages, so the manual is correct.
